runtime/src/main/scala/akka/grpc/internal/TelemetrySpi.scala

```diff
@@ -0,0 +1,59 @@
+package akka.grpc.internal
+
+import akka.actor.{
+  ActorSystem,
+  ClassicActorSystemProvider,
+  ExtendedActorSystem,
+  Extension,
+  ExtensionId,
+  ExtensionIdProvider
+}
+import akka.annotation.{ InternalApi, InternalStableApi }
+import akka.http.scaladsl.model.HttpRequest
+
+import scala.annotation.nowarn
+
+class TelemetryExtensionImpl(val spi: TelemetrySpi) extends Extension
+
+object TelemetryExtension extends ExtensionId[TelemetryExtensionImpl] with ExtensionIdProvider {
+  override def lookup = TelemetryExtension
+  override def createExtension(system: ExtendedActorSystem) =
+    new TelemetryExtensionImpl(TelemetrySpi(system))
+
+  /** Java API */
+  override def get(system: ActorSystem): TelemetryExtensionImpl = super.get(system)
+  override def get(system: ClassicActorSystemProvider): TelemetryExtensionImpl = super.get(system)
+}
+
+private[internal] object TelemetrySpi {
+  private val ConfigKey = "akka.grpc.telemetry-class"
+  def apply(system: ClassicActorSystemProvider): TelemetrySpi = {
+    if (!system.classicSystem.settings.config.hasPath(ConfigKey)) NoOpTelemetry
+    else {
+      val fqcn = system.classicSystem.settings.config.getString(ConfigKey)
+      try {
+        system.classicSystem
+          .asInstanceOf[ExtendedActorSystem]
+          .dynamicAccess
+          .createInstanceFor[TelemetrySpi](fqcn, Nil)
+          .get
+      } catch {
+        case ex: Throwable =>
+          system.classicSystem.log.debug(
+            "{} references a class that could not be instantiated ({}) falling back to no-op implementation",
+            fqcn,
+            ex.toString)
+          NoOpTelemetry
+      }
+    }
+  }
+}
+
+@InternalStableApi
+trait TelemetrySpi {
+  @nowarn
+  def onRequest(prefix: String, method: String, request: HttpRequest): Unit = {}
```

Comment:
May need to be able to modify request and return a new one to pass context. You'll see later.

Reply:
:+1: I suspect not for *this* feature, but indeed would be a nice generalization